### PR TITLE
ADBDEV-6593: Fix vacuum_progress isolation tests

### PR DESCRIPTION
**Pull request by silent-observer**
Fix vacuum_progress isolation tests

Isolation tests `vacuum_progress_column` and `vacuum_progress_row` would
previously fail sometimes because `vacuum_worker_changed` fault injector was
never reached. This happened because vacuum execution and FTS version upgrade
are not synchronized: after FTS disables syncrep according to the test, it
should bump FTS version, which should be detected by the dispatcher on the
next `StartTransaction` call, and then the segment backends will be restarted
(making `vacuum_worker_changed` fault accessible). However, if FTS is too slow
(or dispatcher too fast), it might skip `StartTransaction` calls before FTS
version is bumped, and the segment restart will happen after vacuum is
complete, without triggering `vacuum_worker_changed`. This is an acceptable
behavior for the code since the restart still happens milliseconds after
disabling syncrep, but the test doesn't expect it.

This patch fixes the test by making the dispatcher wait on
`vacuum_rel_finished_one_relation` fault until waits until mirror down is
detected by FTS, ensuring FTS version bump happens before dispatcher proceeds.

Note that the patch also changes expected output: it seems like previously
restart happened on the last `StartTransaction`, making all the collected
stats to be zero. However, post-cleanup vacuum phase (which is executed after
`vacuum_worker_changed`) is also responsible for vacuuming the indexes, so the
correct output should include those too. Note that `num_dead_tuples` value is
still zero, indicating that `vacrelstats` was indeed reset correctly.

### CODE REVIEW

File: src/test/isolation2/sql/vacuum_progress_row.sql [hunk 1]
```diff
@@ -143,8 +143,17 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
...
+-- we shouldn't proceed until FTS probe has bumped the version
+2: SELECT gp_inject_fault('vacuum_rel_finished_one_relation', 'suspend', '', '', 'vacuum_progress_ao_row', 1, 1, 0, 1) FROM master();
+2: SELECT gp_inject_fault('ftsLoop_after_probe', 'skip', 1) FROM master();
 -- resume walsender and let it exit so that mirror stop can be detected
 2: SELECT gp_inject_fault_infinite('wal_sender_loop', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' and content = 1;
+-- wait for FTS probe
+2: SELECT gp_wait_until_triggered_fault('ftsLoop_after_probe', 1, 1) FROM master();
+2: SELECT gp_inject_fault('vacuum_rel_finished_one_relation', 'reset', 1) FROM master();
+2: SELECT gp_inject_fault('ftsLoop_after_probe', 'reset', 1) FROM master();
+
```
Comment (KnightMurloc) [inline src/test/isolation2/sql/vacuum_progress_row.sql]: Can we use `gp_request_fts_probe_scan()` instead of `gp_inject_fault`? It seems to do exactly what we need. Starts the fts probe and waits for it to complete.

**Comment by dkovalev1** [comment]
Could you elaborate a bit more on " This is an acceptable
behavior for the code since the restart still happens milliseconds after
disabling syncrep, but the test doesn't expect it." ?

Why is it acceptable in production and what exactly will happen with DB in prod in this case?

**Comment by dkovalev1** [comment]
Could you also explain some more in " it might skip StartTransaction calls before FTS
version is bumped" ? Who exactly might skip StartTransaction? What will be consequences?

**Comment by silent-observer** [comment]
> Why is it acceptable in production and what exactly will happen with DB in prod in this case?

As far as I know, nothing. Syncrep is disabled when it is detected that mirror is down, so the segment's backend can proceed for a bit before it is restarted, with syncrep turned off. Since mirror being down shouldn't affect the behavior of the segment, if it manages to complete any transactions in that time they should still be correct, so it's acceptable for the restart to happen a bit later.

> Who exactly might skip StartTransaction? What will be consequences?

The dispatcher checks whether the segments should be restarted inside StartTransaction calls. If the dispatcher proceeds through StartTransaction before FTS version is bumped, it will only detect the bump on the *next* StartTransaction. I think there won't be any consequences since the mirror could have been turned off *during* that next transaction as well, with the same observable behavior for the dispatcher. Lack of synchronization only changes the moment when mirror becomes down as observed by the dispatcher, so it shouldn't affect the resulting behavior.

**Review by dkovalev1 — DISMISSED** [review]

File: src/backend/replication/gp_replication.c [hunk 2]
```diff
@@ -587,7 +587,11 @@ GetMirrorStatus(FtsResponse *response, bool *ready_for_syncrep)
 	LWLockRelease(SyncRepLock);
 
 	if (!response->IsMirrorUp)
+	{
 		response->RequestRetry = is_probe_retry_needed();
+		if (!response->RequestRetry)
+			SIMPLE_FAULT_INJECTOR("replication_mirror_down");
+	}
```
Comment (KnightMurloc) [inline src/backend/replication/gp_replication.c]: Maybe we can use `wait_for_mirror_down` (see setup.sql) instead of creating a new fault injector? To wait for the FTS to mark the mirror as down.

Comment (silent-observer) [reply src/backend/replication/gp_replication.c]: Seems to be working, changed to use `wait_for_mirror_down`

**Review by KnightMurloc — DISMISSED** [review]

File: src/test/isolation2/expected/vacuum_progress_column.out [hunk 3]
```diff
@@ -341,12 +341,31 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
  Success:        
  Success:        
 (3 rows)
+
+-- we shouldn't proceed until FTS probe has bumped the version
```
Comment (dkovalev1) [inline src/test/isolation2/expected/vacuum_progress_column.out]: I am afraid this comment seems to be a bit misleading. Perhaps you mean status_version in  FtsVersion in FtsProbeInfo, may be it's better to specify it.
Also, the comment says about bumping the version, but nothing with the code below related to it, instead it waits for the mirror to be down. I suggest to rephrase this comment to match the code.

File: src/test/isolation2/expected/vacuum_progress_column.out [hunk 4]
```diff
@@ -341,12 +341,31 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
...
+ Success:        
+(1 row)
 -- resume walsender and let it exit so that mirror stop can be detected
 2: SELECT gp_inject_fault_infinite('wal_sender_loop', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' and content = 1;
  gp_inject_fault_infinite 
 --------------------------
  Success:                 
 (1 row)
+-- wait for the mirror stop to be detected (timeout 2 minutes)
+2: SELECT wait_for_mirror_down(1::smallint, 120);
```
Comment (dkovalev1) [inline src/test/isolation2/expected/vacuum_progress_column.out:358]: wait_for_mirror_down() initiates extra FTS probe loop between nodes, may be not one, while injected fault used previously used existing notifications, making less disturbance to the system under test, by the cost of extra code change. I am not sure which one is better.

Comment (silent-observer) [reply src/test/isolation2/expected/vacuum_progress_column.out:358]: I think in this case extra FTS probes should be acceptable since all that part of the test is checking is that the backends will eventually be restarted. That way multiple FTS probes shouldn't affect the behavior since they don't bump status_version

File: src/test/isolation2/sql/vacuum_progress_row.sql [hunk 5]
```diff
@@ -143,8 +143,16 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
 -- Resume walsender to detect mirror down and suspend at the beginning
 -- of post-cleanup taken over by a new vacuum worker.
 2: SELECT gp_inject_fault('vacuum_worker_changed', 'suspend', dbid) FROM gp_segment_configuration WHERE content > -1 AND role = 'p';
+
+-- Dispatcher shouldn't proceed until status_version in FtsVersion in FtsProbeInfo is updated,
```
Comment (KnightMurloc) [inline src/test/isolation2/sql/vacuum_progress_row.sql]: Space. In another file the same.
```suggestion
-- Dispatcher shouldn't proceed until status_version in FtsVersion in FtsProbeInfo is updated,
```

**Review by dkovalev1 — DISMISSED** [review]

**Review by KnightMurloc — DISMISSED** [review]

File: src/test/isolation2/sql/vacuum_progress_row.sql [hunk 6]
```diff
@@ -136,6 +136,9 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
 
 -- Resume execution of compact phase and block at syncrep.
 2: SELECT gp_inject_fault_infinite('wal_sender_loop', 'suspend', dbid) FROM gp_segment_configuration WHERE role = 'p' and content = 1;
+-- Dispatcher shouldn't proceed until status_version in FtsVersion in FtsProbeInfo is updated,
+-- ensuring the backends are restarted before they start the post-cleanup phase.
+2: SELECT gp_inject_fault('vacuum_rel_finished_one_relation', 'suspend', '', '', 'vacuum_progress_ao_row', 1, 1, 0, 1) FROM master();
 2: SELECT gp_inject_fault('vacuum_ao_after_compact', 'reset', dbid) FROM gp_segment_configuration WHERE content > -1 AND role = 'p';
```
Comment (KnightMurloc) [inline src/test/isolation2/sql/vacuum_progress_row.sql:142]: Why not reset after we find that the mirror is down?

Comment (silent-observer) [reply src/test/isolation2/sql/vacuum_progress_row.sql:142]: That would change the original behavior of the test: the segments block on syncrep after `vacuum_ao_after_compact` but before `vacuum_rel_finished_one_relation`. If we reset `vacuum_ao_after_compact` after mirror is down, the segment will immediately skip syncrep. I think it's better to suspend on `vacuum_rel_finished_one_relation` since it's positioned roughly at the point where the segments are expected to be without the synchronization (although this was flaky)

File: src/test/isolation2/sql/vacuum_progress_row.sql [hunk 7]
```diff
@@ -136,15 +136,23 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
...
 -- stop the mirror should turn off syncrep
 2: SELECT pg_ctl(datadir, 'stop', 'immediate') FROM gp_segment_configuration WHERE content=1 AND role = 'm';
 
 -- Resume walsender to detect mirror down and suspend at the beginning
 -- of post-cleanup taken over by a new vacuum worker.
 2: SELECT gp_inject_fault('vacuum_worker_changed', 'suspend', dbid) FROM gp_segment_configuration WHERE content > -1 AND role = 'p';
+
```
Comment (KnightMurloc) [inline src/test/isolation2/sql/vacuum_progress_row.sql]: Remove the empty line. It wasn't here before.

**Comment by KnightMurloc** [comment]
A new flaky test?

**Comment by silent-observer** [comment]
Seems so, created a ticket for it

**Review by KnightMurloc — APPROVED** [review]

File: src/test/isolation2/sql/vacuum_progress_column.sql [hunk 8]
```diff
@@ -132,6 +132,9 @@ select relid::regclass as relname, phase, heap_blks_total, heap_blks_scanned, he
 
 -- Resume execution of compact phase and block at syncrep on one segment.
 2: SELECT gp_inject_fault_infinite('wal_sender_loop', 'suspend', dbid) FROM gp_segment_configuration WHERE role = 'p' and content = 1;
+-- Dispatcher shouldn't proceed until status_version in FtsVersion in FtsProbeInfo is updated,
+-- ensuring the backends are restarted before they start the post-cleanup phase.
+2: SELECT gp_inject_fault('vacuum_rel_finished_one_relation', 'suspend', '', '', 'vacuum_progress_ao_column', 1, 1, 0, 1) FROM master();
```
Comment (dkovalev1) [inline src/test/isolation2/sql/vacuum_progress_column.sql:137]: It's not quite clear how moving this fault setup helps in fixing flaky test.

Comment (silent-observer) [reply src/test/isolation2/sql/vacuum_progress_column.sql:137]: The test was flaky because the vacuum managed to clean up indexes before the worker was restarted after FTS version bump. Setting up the fault earlier should help with the edge cases like that.

Another possible solution would be simply removing `heap_blks_scanned` and `heap_blks_vacuumed` columns from the output, since they are mostly irrelevant (the comment mentions num_dead_tuples being reset to 0 after the vacuum worker change, which should be enough to verify the correctness of the test).

Comment (dkovalev1) [reply src/test/isolation2/sql/vacuum_progress_column.sql:137]: Still can't understand, how we now make sure that "vacuum managed to clean up indexes before the worker was restarted after FTS version bump" not happens, and how moving fault helps in it.

Comment (silent-observer) [reply src/test/isolation2/sql/vacuum_progress_column.sql:137]: If the fault is set up earlier, it will stop the dispatcher from proceeding (and cleaning up the indexes). `wal_sender_loop` fault was supposed to take care of that, but in rare circumstances it seems like it doesn't. If we move this `vacuum_rel_finished_one_relation` fault, dispatcher definitely can't proceed.

Comment (dkovalev1) [reply src/test/isolation2/sql/vacuum_progress_column.sql:137]: okay, thanks

**Review by dkovalev1 — APPROVED** [review]
